pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,290 @@
+            X = xd.reshape(eles.ndims, -1)
+            for j, x in enumerate('xyz'[:eles.ndims]):
+                mesh_n[dpf + f"coordsets/coords/values/{x}"] = X[j,:]
+
+            for field in self._expr:
+                mesh_n[dpf + f'fields/{field}/association'] = 'vertex'
+                mesh_n[dpf + f'fields/{field}/volume_dependent'] = 'false'
+                mesh_n[dpf + f'fields/{field}/topology'] = 'mesh'
+
+        self.mesh_n = mesh_n
+
+        # Inital Ascent setup
+        self.a = Ascent()
+        self.ascent_config = ConduitNode()
+        self.ascent_config['mpi_comm'] = comm.py2f()
+        self.ascent_config['runtime/type'] = 'ascent'
+        self.ascent_config['runtine/vtkm/backend'] = 'serial'
+
+        # Open ascent
+        self.a.open(self.ascent_config)
+
+    def __del__(self):
+        self.a.close()
+
+    def _config_scene(self, cfgsect, cons):
+        scene_cfg = {}
+        self.scenes = ConduitNode()
+
+        i = -1 # scene counter
+        for k in self.cfg.items(cfgsect):
+            if k.startswith('scene-'):
+                i += 1
+                configs = self.cfg.getliteral(cfgsect, k)
+                field = configs['field']
+                sname = k.removeprefix('scene-')
+
+                expr = self._get_expr(configs['field_expr'], subs=cons)
+                
+                self._expr[field] = expr
+                scene_cfg[sname] = configs
+
+                for kc in configs:
+                    if kc == 'image_name' or kc == 'field_expr':
+                        pass
+                    elif kc == 'field':
+                        self.scenes[f's{i}/plots/p1/{kc}'] = field
+                    elif kc.startswith('render-'):
+                        ropt = kc.removeprefix('render-')
+                        self.scenes[f's{i}/renders/r1/{ropt}'] = configs[kc]
+                    else:
+                        self.scenes[f's{i}/plots/p1/{kc}'] = configs[kc]
+            
+        return scene_cfg
+
+    def _config_pipeline(self, cfgsect):
+        pipel_cfg = {}
+        self.pipelines = pl = ConduitNode()
+
+        i = -1 # pipeline counter
+        for k in self.cfg.items(cfgsect):
+            if k.startswith('pipeline-'):
+                i += 1
+                configs = self.cfg.getliteral(cfgsect, k)
+                configs = [configs] if isinstance(configs, dict) else configs
+                pname = k.removeprefix('pipeline-')
+
+                pipel_cfg[pname] = configs
+                params = ConduitNode()
+                for j, filter in enumerate(configs):
+                    for kf in filter:
+                        if kf == 'type':
+                            pl[f'pl{i}/f{j}/{kf}'] = filter[kf]
+                        else:
+                            params[f'{kf}'] = filter[kf]
+
+                        if kf == 'field' and kf not in self._expr:
+                            raise KeyError(f"Missing field '{kf}' used in pipeline")
```

Comment:
Missing field 'field'?

Reply:
Error message for when a pipeline expects a named field but it hasn't been defined.

Reply:
Error message just reads weird; we know `kf == 'field'` so can the error message not be constant?